projects/driver_display/src/driver_display_brightness.c

```diff
@@ -0,0 +1,80 @@
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "crc32.h"
+#include "gpio.h"
+#include "log.h"
+#include "misc.h"
+#include "persist.h"
+#include "soft_timer.h"
+#include "status.h"
+
+#include "driver_display_brightness.h"
+#include "driver_display_config.h"
+
+static PersistStorage s_persist;
+static DriverDisplayBrightnessData data;
+
+void driver_display_brightness_timer_callback(SoftTimerID timer_id, void *context) {
+  flash_erase(DRIVER_DISPLAY_CONFIG_PERSIST_PAGE);
+  persist_init(&s_persist, DRIVER_DISPLAY_CONFIG_PERSIST_PAGE, &data, sizeof(data));
+}
+
+void driver_display_brightness_calibration(GPIOAddress adc_address) {
+  // Configures the lower and upper bounds of the given photodiode
+  // Needed to calculate percentage (current brightness value / max brightness value)
+  // The percentage is then used in pwm_set_dc which takes in a percentage value
+  ADCChannel adc_channel;
+  adc_get_channel(adc_address, &adc_channel);
+  LOG_DEBUG("starting config mode for %d seconds", DRIVER_DISPLAY_CONFIG_CALIBRATION_TIME);
+
+  uint16_t reading;
+  soft_timer_start_seconds(DRIVER_DISPLAY_CONFIG_REFRESH_PERIOD,
+                           driver_display_brightness_timer_callback, (void *)&reading,
+                           DRIVER_DISPLAY_CONFIG_REFRESH_TIMER);
+  data.persistStorage = &s_persist;
+
+  // Loop for certain period of time using soft timer
+  while (soft_timer_remaining_time(0) != 0) {
```

Comment:
`soft_timer_remaining_time` is not intended to be used to block and has undefined behavior for completed soft timers.